homeassistant/components/usb/__init__.py

```diff
@@ -0,0 +1,137 @@
+        """Init USB Discovery."""
+        self.hass = hass
+        self.flow_dispatcher = flow_dispatcher
+        self.usb = usb
+        self.seen: set[USBDeviceTupleType] = set()
+
+    async def async_setup(self) -> None:
+        """Set up USB Discovery."""
+        if not await self._async_start_monitor():
+            await self._async_start_scanner()
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STARTED, self.async_start)
+
+    async def async_start(self, event: Event) -> None:
+        """Start USB Discovery and run a manual scan."""
+        self.flow_dispatcher.async_start()
+        await self.hass.async_add_executor_job(self.scan_serial)
+
+    async def _async_start_scanner(self) -> None:
+        """Perodic scan with pyserial when the observer is not available."""
+        stop_track = async_track_time_interval(
+            self.hass, lambda now: self.scan_serial(), SCAN_INTERVAL
+        )
+        self.hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP, callback(lambda event: stop_track())
+        )
+
+    async def _async_start_monitor(self) -> bool:
+        """Start monitoring hardware with pyudev."""
+        if not sys.platform.startswith("linux"):
+            return False
+        from pyudev import (  # pylint: disable=import-outside-toplevel
+            Context,
+            Monitor,
+            MonitorObserver,
+        )
+
+        try:
+            context = Context()
+        except ImportError:
+            return False
+
+        monitor = Monitor.from_netlink(context)
+        monitor.filter_by(subsystem="tty")
+        observer = MonitorObserver(
+            monitor, callback=self._device_discovered, name="usb-observer"
+        )
+        observer.start()
+        self.hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP, lambda event: observer.stop()
+        )
+        return True
+
+    def _device_discovered(self, device):
+        """Call when the observer discovers a new usb tty device."""
+        if device.action != "add":
+            return
+        _LOGGER.debug(
+            "Discovered Device at path: %s, triggering scan serial",
+            device.device_path,
+        )
+        self.scan_serial()
+
+    @callback
+    def _async_process_discovered_usb_device(self, device: USBDevice) -> None:
+        """Process a USB discovery."""
+        _LOGGER.debug("Discovered USB Device: %s", device)
+        device_tuple = usb_device_tuple(device)
+        if device_tuple in self.seen:
+            return
+        self.seen.add(device_tuple)
+        for matcher in self.usb:
+            if "vid" in matcher and device["vid"] != matcher["vid"]:
+                continue
+            if "pid" in matcher and device["pid"] != matcher["pid"]:
+                continue
+            flow: USBFlow = {
+                "domain": matcher["domain"],
+                "context": {"source": config_entries.SOURCE_USB},
+                "data": dict(device),
```

Comment:
If we make `USBDevice` a dataclass we can use `dataclasses.asdict` for this.

Reply:
Done